lib/iris/fileformats/_nc_load_rules/actions.py

```diff
@@ -0,0 +1,557 @@
+    """Register a CFVariable as a formula root."""
+    rule_name = "fc_formula_type"
+    (var_name,) = formula_root_fact
+    cf_var = engine.cf_var.cf_group[var_name]
+    # cf_var.standard_name is a formula type (or we should never get here).
+    formula_type = getattr(cf_var, "standard_name", None)
+    succeed = True
+    if formula_type not in iris.fileformats.cf.reference_terms:
+        succeed = False
+        rule_name += f"(FAILED - unrecognised formula type = {formula_type!r})"
+        msg = f"Ignored formula of unrecognised type: {formula_type!r}."
+        warnings.warn(msg)
+    if succeed:
+        # Check we don't already have one.
+        existing_type = engine.requires.get("formula_type")
+        if existing_type:
+            # NOTE: in this case, for now, we will accept the last appearing,
+            # which matches the older behaviour.
+            # TODO: this needs resolving, somehow.
+            succeed = False
+            msg = (
+                "Omitting factories for some hybrid coordinates, as multiple "
+                "hybrid coordinates on a single variable are not supported: "
+                f"Formula of type ={formula_type!r} "
+                f"overrides another of type ={existing_type!r}.)"
+            )
+            warnings.warn(msg)
+        rule_name += f"_{formula_type}"
+        # Set 'requires' info for iris.fileformats.netcdf._load_aux_factory.
+        engine.requires["formula_type"] = formula_type
+
+    return rule_name
+
+
+@action_function
+def action_formula_term(engine, formula_term_fact):
+    """Register a CFVariable as a formula term."""
+    # Must run AFTER formula root identification.
+    (termvar_name, rootvar_name, term_name) = formula_term_fact
+    # The rootname is implicit :  have only one per cube
+    # TODO: change when we adopt cf-1.7 advanced grid-mapping syntax
+    engine.requires.setdefault("formula_terms", {})[term_name] = termvar_name
+    rule_name = f"fc_formula_term({term_name})"
+    return rule_name
+
+
+def run_actions(engine):
+    """
+    Run all actions for a cube.
+
+    This is the top-level "activation" function which runs all the appropriate
+    rules actions to translate facts and build all the cube elements.
+
+    The specific cube being translated is "engine.cube".
+
+    """
+
+    # default (all cubes) action, always runs
+    action_default(engine)  # This should run the default rules.
+
+    # deal with grid-mappings
+    grid_mapping_facts = engine.fact_list("grid_mapping")
+    # For now, there should be at most *one* of these.
+    assert len(grid_mapping_facts) in (0, 1)
+    for grid_mapping_fact in grid_mapping_facts:
+        action_provides_grid_mapping(engine, grid_mapping_fact)
+
+    # identify + record aka "PROVIDE" specific named coordinates
+    # N.B. cf.py has identified that these are dim-coords, NOT aux-coords
+    # (which are recorded separately).
+    # TODO: can probably remove this step ??
```

Comment:
I'm assuming that the appropriate time to remove this step would be when simplifying the new behaviour as opposed to now when we are simply copying the behaviour.

Reply:
Yes, I think so.
It is certainly now _possible_ to further simplify the operations, 
and especially to combine the 'provides-coordinate' and 'build-dimension-coordinate' routines.
But for the present I think it remains simpler+safer  to leave this all as it is, in case it may help us debug any unexpected user-reported cases where the new loading does not match the old.
While the overhead of understanding this code might still be a bit high, the practical inefficiency of it should be negligible.